compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirDelegatedPropertyChecker.kt

```diff
@@ -0,0 +1,113 @@
+import org.jetbrains.kotlin.fir.resolve.diagnostics.ConeInapplicableCandidateError
+import org.jetbrains.kotlin.fir.resolve.diagnostics.ConeUnresolvedNameError
+import org.jetbrains.kotlin.fir.symbols.AbstractFirBasedSymbol
+import org.jetbrains.kotlin.fir.types.ConeKotlinErrorType
+import org.jetbrains.kotlin.fir.types.coneType
+import org.jetbrains.kotlin.fir.types.render
+import org.jetbrains.kotlin.fir.visitors.FirVisitorVoid
+import org.jetbrains.kotlin.resolve.calls.tower.CandidateApplicability
+import org.jetbrains.kotlin.resolve.calls.tower.isSuccess
+
+object FirDelegatedPropertyChecker : FirPropertyChecker() {
+    override fun check(declaration: FirProperty, context: CheckerContext, reporter: DiagnosticReporter) {
+        val delegate = declaration.delegate ?: return
+        val delegateType = delegate.typeRef.coneType
+
+        if (delegateType is ConeKotlinErrorType) {
+            val delegateSource = delegate.source
+            // Implicit recursion type is not reported since the type ref does not have a real source.
+            if (delegateSource != null && (delegateType.diagnostic as? ConeSimpleDiagnostic)?.kind == DiagnosticKind.RecursionInImplicitTypes) {
+                // skip reporting other issues in this case
+                reporter.reportOn(delegateSource, FirErrors.RECURSION_IN_IMPLICIT_TYPES, context)
+            }
+            return
+        }
+
+        class DelegatedPropertyAccessorVisitor(private val isGet: Boolean) : FirVisitorVoid() {
+            override fun visitElement(element: FirElement) = element.acceptChildren(this)
+
+            override fun visitFunctionCall(functionCall: FirFunctionCall) {
+                val errorNamedReference = functionCall.calleeReference as? FirErrorNamedReference ?: return
+                if (errorNamedReference.source?.kind != FirFakeSourceElementKind.DelegatedPropertyAccessor) return
+                val expectedFunctionSignature =
+                    (if (isGet) "getValue" else "setValue") + "(${functionCall.arguments.joinToString(", ") { it.typeRef.coneType.render() }})"
+                val delegateDescription = if (isGet) "delegate" else "delegate for var (read-write property)"
+                fun reportInapplicableDiagnostics(
+                    candidateApplicability: CandidateApplicability,
+                    candidates: Collection<AbstractFirBasedSymbol<*>>
+                ) {
+                    if (candidateApplicability == CandidateApplicability.INAPPLICABLE_WRONG_RECEIVER) {
+                        reporter.reportOn(
+                            errorNamedReference.source,
+                            FirErrors.DELEGATE_SPECIAL_FUNCTION_MISSING,
+                            expectedFunctionSignature,
+                            delegateType,
+                            delegateDescription,
+                            context
+                        )
+                    } else {
```

Comment:
Are you only expecting `CandidateApplicability.UNSAFE_CALL` here? In which case, we can just use a `when (candidateApplicability) ...`

Reply:
I am not sure I understand. Do you mean to use `when` instead of `if` here?

Reply:
I was suggesting that if you were only expecting `UNSAFE_CALL` in the `else` branch, then we can do
```kt
when (candidateApplicability) {
  CandidateApplicability.INAPPLICABLE_WRONG_RECEIVER -> { ... }
  CandidateApplicability.UNSAFE_CALL -> { ... }
```
... so that we don't report anything unexpectedly on some other `CandidateApplicability`. So my first question was, are we expecting anything else aside from `UNSAFE_CALL`?

Reply:
Yes, we are expecting errors other than `UNSAFE_CALL`. In fact, any applicability is possible here.

Reply:
Ah, of course. Got it now. The `if-else` was more appropriate then, sorry about that.